lib/internal/modules/esm/loader.js

```diff
@@ -408,41 +481,103 @@ class ESMLoader {
 
     const namespaces = await PromiseAll(new SafeArrayIterator(jobs));
 
-    return wasArr ?
-      namespaces :
-      namespaces[0];
+    if (!wasArr) { return namespaces[0]; }
+
+    for (let i = 0; i < count; i++) {
+      const namespace = ObjectCreate(null);
+      namespace.url = specifiers[i];
+      namespace.exports = namespaces[i];
+
+      namespaces[i] = namespace;
+    }
+
+    return namespaces;
   }
 
   /**
    * Provide source that is understood by one of Node's translators.
    *
-   * The internals of this WILL change when chaining is implemented,
-   * depending on the resolution/consensus from #36954
-   * @param {string} url The URL/path of the module to be loaded
+   * Internally, this behaves like a backwards iterator, wherein the stack of
+   * hooks starts at the top and each call to `next()` moves down 1 step until
+   * it reaches the bottom or short-circuits.
+   *
+   * @param {URL['href']} url The URL/path of the module to be loaded
    * @param {object} context Metadata about the module
-   * @returns {object}
+   * @returns {{ format: ModuleFormat, source: ModuleSource }}
    */
   async load(url, context = {}) {
-    const defaultLoader = this.#loaders[0];
+    const loaders = this.#loaders;
+    let hookIndex = loaders.length - 1;
+    let {
+      fn: loader,
+      url: loaderFilePath,
+    } = loaders[hookIndex];
+    let chainFinished = hookIndex === 0;
+
+    const next = (nextUrl, ctx = context) => {
+      // `next` has been called, so decrement our pointer
+      --hookIndex;
+
+      ({
+        fn: loader,
+        url: loaderFilePath,
+      } = loaders[hookIndex]);
+
+      if (hookIndex === 0) { chainFinished = true; }
+
+      const hookErrIdentifier = `${loaderFilePath} "load"`;
+
+      if (typeof nextUrl !== 'string') {
+        // non-strings can be coerced to a url string
+        // validateString() throws a less-specific error
+        throw new ERR_INVALID_ARG_TYPE(
+          `${hookErrIdentifier} next(url)`,
+          'a url string',
+          nextUrl,
+        );
+      }
 
-    const loader = this.#loaders.length === 1 ?
-      defaultLoader :
-      this.#loaders[1];
-    const loaded = await loader(url, context, defaultLoader);
+      try {
+        new URL(nextUrl);
+      } catch {
+        throw new ERR_INVALID_ARG_VALUE(
+          `${hookErrIdentifier} next(url)`,
+          nextUrl,
+          'should be a url string',
+        );
+      }
```

Comment:
One way to avoid repeated `URL` calls could just be to check if the `nextUrl` is in the current loader registry. Since all registry URLs are known valid URLs that provides a simple cached check for perf.